podtato-head-monolith/podtatoserver.go

```diff
@@ -0,0 +1,147 @@
+var getCallCounter = prometheus.NewCounterVec(
+	prometheus.CounterOpts{
+		Name: "http_requests_total", // metric name
+		Help: "Number of get requests.",
+	},
+	[]string{"status"}, // labels
+)
+
+var buckets = []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10}
+
+var responseTimeHistogram = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	Name:    "http_server_request_duration_seconds",
+	Help:    "Histogram of response time for handler in seconds",
+	Buckets: buckets,
+}, []string{"route", "method", "status_code"})
+
+// create a handler struct
+type HTTPHandler struct{}
+
+type statusRecorder struct {
+	http.ResponseWriter
+	statusCode int
+}
+
+func (rec *statusRecorder) WriteHeader(statusCode int) {
+	rec.statusCode = statusCode
+	rec.ResponseWriter.WriteHeader(statusCode)
+}
+
+func getRoutePattern(r *http.Request) string {
+	reqContext := mux.CurrentRoute(r)
+	if pattern, _ := reqContext.GetPathTemplate(); pattern != "" {
+		return pattern
+	}
+
+	fmt.Println(reqContext.GetPathRegexp())
+
+	return "undefined"
+}
+
+// implement `ServeHTTP` method on `HttpHandler` struct
+func (h HTTPHandler) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	var status string
+	defer func() {
+		// increment the counter on defer func
+		getCallCounter.WithLabelValues(status).Inc()
+	}()
+
+	overviewTemplate = template.Must(template.ParseFiles("./static/podtato-new.html"))
+	err := overviewTemplate.Execute(res, podtatoConfiguration)
+
+	if err != nil {
+		log.Print(err.Error())
+	}
+	// Slow build
+	if podtatoConfiguration.ServiceVersion == "0.1.2" {
+		time.Sleep(2 * time.Second)
+	}
+
+	if err != nil {
+		status = "error"
+	}
+	status = "success"
+}
+
+func prometheusMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		rec := statusRecorder{w, 200}
+
+		next.ServeHTTP(&rec, r)
+
+		duration := time.Since(start)
+		statusCode := strconv.Itoa(rec.statusCode)
+		route := getRoutePattern(r)
+		fmt.Println(duration.Seconds())
+		responseTimeHistogram.WithLabelValues(route, r.Method, statusCode).Observe(duration.Seconds())
+	})
+}
+
+func init() {
+	prometheus.Register(getCallCounter)
+	prometheus.Register(responseTimeHistogram)
+}
+
+func main() {
+	// expecting version as first parameter
+	serviceVersion := flag.String("version", "", "Service version e.g. 0.1.0")
+	flag.Parse()
+
+	var err error
+	podtatoConfiguration, err = pkg.GetAssembledPodtatoConfiguration(*serviceVersion)
+	if err != nil {
+		log.Fatal(err)
+	}
+	// create a new handler
+	handler := HTTPHandler{}
+
+	router := mux.NewRouter()
+	router.Use(prometheusMiddleware)
+
+	router.Path("/").Handler(handler)
+	router.Path("/")
+
+	staticDir := os.Getenv("STATIC_DIR")
+	if staticDir == "" {
+		staticDir = "/static"
+	}
+
+	// Serving static files
+	router.
+		PathPrefix(staticDir).
+		Handler(http.StripPrefix(staticDir, http.FileServer(http.Dir(staticDir))))
```

Comment:
After testing this out it did not expose the `static` directory with the absolute path (`/static'`) in the `http.Dir` command. Changing it to a relative path (`./static`) worked:

```
Handler(http.StripPrefix(staticDir, http.FileServer(http.Dir("."+staticDir))))
```

Reply:
Interesting! I had problems with the existing default prefix `./static/` and only got the test (`build/test_image.sh`) to work with `/static`. _That_ works cause in the Dockerfile the binary itself and the `static` dir are copied to the root (`/`); it wouldn't work if running the binary directly though.

One problem in the code is that it uses the same string variable to specify the path prefix in the URL _and_ the path to find the files on the file system; let me see about changing that.

This also motivates me to add a proper GitHub workflow to build the image and run the test so we can ensure it works the same for everyone.

Reply:
I added a GitHub workflow to build and test the images using the scripts I added and it passes: <https://github.com/podtato-head/podtato-head/actions/runs/1654869776>.

Not sure why this code doesn't work for you @TannerGabriel! Are you running the binary outside of the image perhaps?

Reply:
A proper Github workflow sounds excellent!

Reply:
Added in latest commit, PTAL!

Reply:
In the ensuing commit I distinguished the file system path from the URL path prefix for static assets. Now the filesystem path is `./static` and the URL path prefix is `/static`, which seems correct. Tests pass. Can you try it @TannerGabriel?

Reply:
Yes, that worked for me!